Implement OrderTester

Ashalynd opened the pull request:
Fixes #568

Ashalynd commented:
By the way, didn't know that forks get detached if the parent repo becomes public :)

Strilanc commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,155 @@
+        self.groups = [(_ClassUnknownToSubjects(),)]
+
+    def _verify_ascending(self, v1, v2):
+        # __lt__ and __gt__ might not necessarily be implemented both.
```
Check that the usual syntax works in all eight cases, which can each invoke different methods in different orders:

```python
assert v1 < v2
assert v1 <= v2
assert v2 > v1
assert v2 >= v1
assert not v1 > v2
assert not v2 >= v1
assert not v1 < v2
assert not v1 <= v2
```

Each of those methods may be implemented with different code (especially in python 2).

Then also check that all four of the `__lt__`/`__le__`/`__ge__`/`__gt__` methods works for both values (returns the correct thing or returns `NotImplemented` or is not present).

Ashalynd replied:
Ok, so you are saying that we should assume that all relational operators are implemented.

Strilanc replied:
We should test that all relational binary operators work, because it is possible that they trigger different code. Testing the explicit methods is less important, and I don't think we should not require that they be present.

Ashalynd replied:
Not-versions create problems for the initial add:
```        assert not hasattr(v1, '__le__') or not v1 <= v2```

Strilanc commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,155 @@
+        gt = hasattr(v2, '__gt__') and v2.__gt__(v1)
+        assert lt or gt
+
+    def _verify_equality(self, v1, v2):
```
Delegate to an `EqualsTester` instance.

Strilanc commented on cirq/testing/order_tester_test.py:
```diff
@@ -0,0 +1,66 @@
+    with pytest.raises(AssertionError):
+        ot.add_ascending_equivalence_group(0, 0.)  # not ascending w.r.t
+                                                   # previous items
+
```
Test cases to include:

InconsistentWithOthers/AlwaysTrue/AlwaysFalse `__le__`/`__ge__`/`__lt__`/`__gt__`.

`__le__` or `__ge__` returns false when `__eq__` returns true.

`__lt__` and `__gt__` both return false when `__eq__` returns false.

Ashalynd replied:
Can this be done in a follow-up commit?

Strilanc replied:
No, this is crucial functionality that must be work from the start.

Strilanc commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,155 @@
+import itertools
+
+
+class OrderTester:
```
Perhaps we should rename to `TotalOrderTester`? Or have a parameter that switches between partial order and total order? I'm not sure what a partial order tester would even look like, other than specifying a graph and checking the transitive completion.

Ashalynd replied:
Let's do it in a separate PR.

Strilanc commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,155 @@
+        # Check that already added groups are not overlapping with the new
+        # group.
+        if self.groups:
+            self._verify_ascending(self.groups[-1][-1], group_items[0])
```
Yes, for a correct implementation this would be sufficient. But an incorrect implementation may have A < B < C and then a bug where C > A.

Strilanc commented:
Another reason to test all the cases is that it covers code. For example, in my NamedQubit change https://github.com/quantumlib/Cirq/pull/833 that relies on this PR I'm getting a coverage failure because the order tester is not trying everything:

```
************* cirq/ops/named_qubit.py (8 uncovered)
Line   32 is new or changed but not covered:              return NotImplemented
Line   39 is new or changed but not covered:          if not isinstance(other, type(self)):
Line   40 is new or changed but not covered:              return NotImplemented
Line   41 is new or changed but not covered:          return _pad_digits(self.name) >= _pad_digits(other.name)
Line   49 is new or changed but not covered:          if not isinstance(other, type(self)):
Line   50 is new or changed but not covered:              return NotImplemented
Line   51 is new or changed but not covered:          return _pad_digits(self.name) <= _pad_digits(other.name)
Line   55 is new or changed but not covered:              return NotImplemented
```

Ashalynd commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,155 @@
+                'Items in the same group produced different hashes. '
+                'Example: hash({}) is {} but hash({}) is {}.'.format(*example))
+
+    def add_ascending(self, *items: Any):
```
Good point, thanks for noticing.

Ashalynd commented:
Addressed most of the comments, except there is still some code uncovered with tests.

Strilanc commented on cirq/testing/equals_tester.py:
```diff
@@ -33,8 +33,8 @@ class EqualsTester:
 
-    def add_equality_group(self, *group_items: Any):
-        """Tries to add a disjoint equivalence group to the equality tester.
+    def verify_equality_group(self, *group_items: Any):
```
Don't bother having this be a separate method; just stick with `add_equlity_group` being the only one and use that.

Or is this intended to be an internal helper? In that case prefix it with `_`.

Ashalynd replied:
I wanted to run base class method (verify equality) prior to running the method from the child class (verify ascending).

Strilanc replied:
Ah, I see. Yes, it does make more sense in the context of a child class trying to avoid creating equivalence groups behind the scenes. But OrderTester shouldn't inherit EqualsTester so that becomes irrelevant.

Strilanc commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,230 @@
+        not_ge_1 = NotImplemented
+
+        try:
+            lt_1 = v1 < v2
```
None of these should ever throw a type error. The user asserted the ordering should be present.

Ashalynd replied:
If you try adding a group with (object, object), that's what actually happens. Though in that case it might be not a valid thing to do in the first place.

Strilanc commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,230 @@
+                ]
+
+        try:
+            gt_2 = not hasattr(v2, '__gt__') or v2 > v1
```
No need to do `hasattr` guards before calling the binary operator. It should just always work, given that the user is asserting the ordering is present.

Strilanc commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,230 @@
+from cirq.testing.equals_tester import EqualsTester
+
+
+class OrderTester(EqualsTester):
```
Don't inherit from EqualsTester. OrderTester is not a superset of EqualsTester, and having the `add_equivalence_group` methods presence seriously muddles the usage. For example, from the outside, what should happen if I alternate between `add_equivalence_group` and `add_ascending_equivalence_group`?

Instead, have an `__init__` method initialize a property `_eq_tester`, and just delegate the equality checks to that.

Strilanc commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,230 @@
+
+    def _verify_ascending(self, v1, v2):
+        """Verifies that (v1, v2) is a strictly ascending sequence."""
+        lt_1 = NotImplemented
```
I *think* this method can just be the eight asserts I listed earlier (four comparisons * two outcomes). The correct `NotImplemented` behavior will be checked by a new helper method `_assert_not_implemented_vs_unknown` that is only called when that should be relevant.

Strilanc commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,230 @@
+        return hash(ClassSmallerThanEverythingElse)  # coverage: ignore
+
+
+class ClassLargerThanEverythingElse:
```
This class should be private (prefix `_`). The caller is not supposed to know about it.

Strilanc commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,230 @@
+        return not isinstance(other, ClassSmallerThanEverythingElse)
+
+    def __hash__(self):
+        return hash(ClassSmallerThanEverythingElse)  # coverage: ignore
```
It shouldn't be necessary to say the coverage of this method is ignored. This method should be indirectly invoked via comparisons within EqualsTester.

Ashalynd replied:
It didn't do so for the tests I had. It might be because in the existing code in EqualsTester, it only tests hashes after verifying that the new elements aren't equal to anything else which has been added before, which would always fail for that class.

Strilanc commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,230 @@
+    def __init__(self):
+        self.groups = [(ClassSmallerThanEverythingElse(),)]
+
+    def _verify_ascending(self, v1, v2):
```
Add an `_assert_not_implemented_vs_unknown(item)` method. In this method call `_verify_ascending(ClassSmallerThanEverythingElse(), item)` and `_verify_ascending(item, ClassLargerThanEverythingElse())`. Finally, use `with pytest.raises(TypeError)` around the eight possible ways to compare against `_Incomparable()`.

Then invoke this method when adding items.

Strilanc replied:
use `_assert` instead of `_verify` naming.

Strilanc commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,230 @@
+    """Tests ordering against user-provided disjoint ordered groups or items."""
+
+    def __init__(self):
+        self.groups = [(ClassSmallerThanEverythingElse(),)]
```
Get rid of the smaller than everything else group, since the check this implied will be handled by the `_assert_not_implemented_vs_unknown` helper method.

Strilanc commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,230 @@
+            (NotImplemented, NotImplemented),
+                ]
+
+    def verify_ascending_group(self, *group_items: Any):
```
I don't think this should be a separate available public method. The user should not be choosing whether to add+verify or just verify a group or not. Just always add+verify.

Merge this back into `add_ascending_equivalence_group` or make it private (prefix with `_`).

Strilanc commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,230 @@
+        self.groups.append(tuple(group_items))
+
+
+class ClassSmallerThanEverythingElse:
```
This class should be private (prefix `_`). The caller is not supposed to know about it.

Strilanc commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,230 @@
+        return hash(ClassLargerThanEverythingElse)
+
+
+class UnorderableClass:
```
This class should be private (prefix `_`). The caller is not supposed to know about it.

Ashalynd replied:
Do we even need it there? It's only used in the tests. Ditto about two other classes.

Strilanc replied:
This class is supposed be used as a verification step within OrderTester. The tests should not refer to it directly, but they should have example classes that would fail due to these things being checked.

Strilanc commented on cirq/testing/order_tester_test.py:
```diff
@@ -0,0 +1,84 @@
+        def __hash__(self):
+            return self._h
+
+
```
Add test cases for common mistakes made when implement binary operators, such as forgetting to return `NotImplemented` when seeing an unknown type or accidentally implementing `__gt__` instead of `__ge__`.

Ashalynd replied:
Didn't address that yet. May be do it in a follow-up?

Strilanc commented on cirq/testing/equals_tester.py:
```diff
@@ -61,8 +61,13 @@ def add_equality_group(self, *group_items: Any):
+                try:
+                    eq = v1.__eq__(v2)
+                    ne = v1.__ne__(v2)
+                except TypeError:
```
This should never ever be happening! Don't ignore this error, it is an actual failure!

Strilanc commented on cirq/testing/equals_tester.py:
```diff
@@ -78,8 +83,13 @@ def add_equality_group(self, *group_items: Any):
+                    try:
+                        eq = v1.__eq__(v2)
+                        ne = v1.__ne__(v2)
+                    except TypeError:
```
This should never ever be happening! Don't ignore this error, it is an actual failure!

Ashalynd replied:
Actually it only happens if one tries to add sequence of something like (object, object).
<pre>&gt;&gt;&gt; a = object
&gt;&gt;&gt; b = object
&gt;&gt;&gt; a &lt; b
Traceback (most recent call last):
  File &quot;&lt;stdin&gt;&quot;, line 1, in &lt;module&gt;
TypeError: unorderable types: type() &lt; type()
</pre>

Ashalynd replied:
Digged up the actual error from the logs before that change:

```
______________________ test_add_ordering_group_incorrect _______________________

    def test_add_ordering_group_incorrect():
        ot = OrderTester()
        ot.add_ascending(1, 2)
        with pytest.raises(AssertionError):
>           ot.add_ascending(object, object)  # not ascending within call

cirq/testing/order_tester_test.py:31: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cirq/testing/order_tester.py:148: in add_ascending
    self.add_ascending_equivalence_group(item)
cirq/testing/order_tester.py:166: in add_ascending_equivalence_group
    super(OrderTester, self).verify_equality_group(*group_items)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <cirq.testing.order_tester.OrderTester object at 0x7fa71a6ca128>
group_items = (<class 'object'>,), v1 = <class 'object'>, v2 = <class 'object'>

    def verify_equality_group(self, *group_items: Any):
        """Verifies that a group is an equivalence group.
    
            This methods asserts that items within the group must all be equal to
            each other, but not equal to any items in other groups that have been
            or will be added.
    
            Args:
              *group_items: The items making up the equivalence group.
    
            Raises:
                AssertionError: Items within the group are not equal to each other,
                    or items in another group are equal to items within the new
                    group, or the items violate the equals-implies-same-hash rule.
            """
    
        assert group_items
    
        # Check that group items are equivalent to each other.
        for v1, v2 in itertools.product(group_items, group_items):
            # Binary operators should always work.
            assert v1 == v2
            assert not v1 != v2
    
            # __eq__ and __neq__ should both be correct or not implemented.
            assert hasattr(v1, '__eq__') == hasattr(v1, '__ne__')
            # Careful: python2 int doesn't have __eq__ or __ne__.
            if hasattr(v1, '__eq__'):
>               eq = v1.__eq__(v2)
E               TypeError: expected 1 arguments, got 0

cirq/testing/equals_tester.py:64: TypeError
```

Ashalynd replied:
That seems to be a known issue for class objects: https://bugs.python.org/issue10017

Strilanc replied:
It looks like this is a perfect reason not to use the `__eq__` and friends directly as I've been doing so far in the tests. I've fixed `EqualsTester` to only use `==` and `!=` instead of worrying about `__eq__`, and we should the same here.

Strilanc commented:
Here is a reasonable way to assert that items are ascending. I will fix the type error bug in EqualsTester when adding `object`.

```python
    @staticmethod
    def _assert_ascending(v1: Any, v2: Any) -> None:
        comparisons = [
            (lambda a, b: a < b, '<'),
            (lambda a, b: a > b, '>'),
            (lambda a, b: a <= b, '<='),
            (lambda a, b: a >= b, '>='),
        ]

        for x, y, ascend in [(v1, v2, True), (v2, v1, False)]:
            for cmp, label in comparisons:
                actual = cmp(x, y)
                expect = cmp(0, 1 if ascend else -1)
                assert actual == expect, (
                    "Expected {!r} to come {} {!r}, "
                    "but '{!r} {} {!r}' evaluated to {}".format(
                        x,
                        'before' if ascend else 'after',
                        y,
                        x,
                        label,
                        y,
                        actual))
```

Strilanc commented:
https://github.com/quantumlib/Cirq/pull/838 fixes the `object` TypeError thing in EqualsTester

Strilanc commented:
I'm on vacation for a couple weeks, so I'm handing this off to @dabacon

Strilanc commented on cirq/testing/equals_tester.py:
```diff
@@ -53,46 +62,19 @@ def verify_equality_group(self, *group_items: Any):
-                                    (NotImplemented, NotImplemented)]
-
-        # Check that this group's items don't overlap with other groups.
+            same = EqualsTester._eq_check(v1, v2)
```
Make an attribute `self._equals_tester = EqualsTester()` and add the groups to it as they come in.

Strilanc replied:
Oops, was confused about where this was!

Ashalynd commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,230 @@
+        except TypeError:
+            pass
+
+        assert(lt_1, not_lt_2) in [
```
Ahhh... This has been failing in the Python 2 version we use (2.7.6) due to the regression which was fixed in Python 2.7.8: https://mail.python.org/pipermail/python-checkins/2014-May/130168.html

Do we want to stick to Python 2.7.6 or to push it to the latest Python 2.7 version (2.7.14)? The issue mentioned above doesn't happen in this version either. If not, then I can change the condition, because one can also argue that people could have whatever version of Python 2.7 :)

Strilanc replied:
I think we should be compatible with the python version that is on travis.

Ashalynd commented:
OK... didn't realize that in Python 2, the numerics don't have rich comparison operators.

```
Python 2.7.13 (default, Nov 24 2017, 17:33:09) 
>>> a = 4
>>> hasattr(a, '__gt__')
False
>>> dir(a)
['__abs__', '__add__', '__and__', '__class__', '__cmp__', '__coerce__', '__delattr__', '__div__', '__divmod__', '__doc__', '__float__', '__floordiv__', '__format__', '__getattribute__', '__getnewargs__', '__hash__', '__hex__', '__index__', '__init__', '__int__', '__invert__', '__long__', '__lshift__', '__mod__', '__mul__', '__neg__', '__new__', '__nonzero__', '__oct__', '__or__', '__pos__', '__pow__', '__radd__', '__rand__', '__rdiv__', '__rdivmod__', '__reduce__', '__reduce_ex__', '__repr__', '__rfloordiv__', '__rlshift__', '__rmod__', '__rmul__', '__ror__', '__rpow__', '__rrshift__', '__rshift__', '__rsub__', '__rtruediv__', '__rxor__', '__setattr__', '__sizeof__', '__str__', '__sub__', '__subclasshook__', '__truediv__', '__trunc__', '__xor__', 'bit_length', 'conjugate', 'denominator', 'imag', 'numerator', 'real']
>>> quit()

Python 3.5.3 (default, Jan 19 2017, 14:11:04) 
>>> a = 4
>>> dir(a)
['__abs__', '__add__', '__and__', '__bool__', '__ceil__', '__class__', '__delattr__', '__dir__', '__divmod__', '__doc__', '__eq__', '__float__', '__floor__', '__floordiv__', '__format__', '__ge__', '__getattribute__', '__getnewargs__', '__gt__', '__hash__', '__index__', '__init__', '__int__', '__invert__', '__le__', '__lshift__', '__lt__', '__mod__', '__mul__', '__ne__', '__neg__', '__new__', '__or__', '__pos__', '__pow__', '__radd__', '__rand__', '__rdivmod__', '__reduce__', '__reduce_ex__', '__repr__', '__rfloordiv__', '__rlshift__', '__rmod__', '__rmul__', '__ror__', '__round__', '__rpow__', '__rrshift__', '__rshift__', '__rsub__', '__rtruediv__', '__rxor__', '__setattr__', '__sizeof__', '__str__', '__sub__', '__subclasshook__', '__truediv__', '__trunc__', '__xor__', 'bit_length', 'conjugate', 'denominator', 'from_bytes', 'imag', 'numerator', 'real', 'to_bytes']
```

Strilanc commented on cirq/testing/order_tester.py:
```diff
@@ -24,78 +24,65 @@
+        lt_1 = self._try_comparison(lambda: v1 < v2)
+        not_lt_2 = self._try_comparison(lambda: not v2 < v1)
+
+        self._do_assert(lt_1, not_lt_2, "{!r}__lt__{!r}".format(v1, v2))
```
This is too lenient. The check should fail if the values are incomparible (i.e. throw a type error). Just `assert v1 < v2` and `assert not v2 < v1`.

Ashalynd replied:
The current checks in that method are supposed to do the following (assuming we want to verify ascending order for a, b):

- for Python3, verify that strictly either a < b or b > a; also verify that a<=b,  b>=a or le / ge arent defined

- for Python2, only verify that either a<b or b>a.

But right, type error should just mean a no go. I  thought that in theory it could be possible to have type error only for some comparison options, but probably it's too exotic.

Strilanc replied:
We want both comparisons to succeed in python 2 also. Why would we be okay with a class that failed if you did b>a instead of a<b?

Ashalynd commented:
Addressed a few comments:

- removed the dependency on EqualsTester
- treat Python 2 and Python 3 in the same way
- do not eat TypeError (rethrow as AssertionError)

Strilanc commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,223 @@
+    def _verify_ascending(self, v1, v2):
+        """Verifies that (v1, v2) is a strictly ascending sequence."""
+
+        lt_1 =  self._try_comparison(lambda: v1 < v2)
```
Define a list of the six comparison operators with names and lambdas at the start of the file, then iterate through them here. Once with `v1` as the first item, and once with `v2` as the first item. Assert that the result on v1 vs v2 should match the result on int 1 vs int 2.

```python
comparisons = [
    ('<', lambda a, b: a < b),
    ...
]
```

```python
for s, f in comparisons:
    assert f(v1, v2) == f(1, 2), '{} {} {} returned {} instead of {}'.format(v1, s, v2, f(v1, v2), f(1, 2))
```

Ashalynd replied:
You mean four?
<, >, <=, >=

Strilanc commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,223 @@
+
+    def __init__(self):
+        self.eq_tester = EqualsTester()
+        self.eq_tester.groups = []  # otherwise ClassUnknownToSubjects
```
Keep track of your own groups instead of using equal tester's, so these kinds of hacks aren't needed.

Should probably make the groups variable private actually...

Ashalynd replied:
EqualsTester relies on its own internal state (self.groups) when it verifies equality. This could be solved by providing it an optional groups to be used instead of self.groups. I would rather do it later.

Strilanc commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,223 @@
+        try:
+            result = comp()
+        except TypeError as inst:
+            assert False, inst
```
Not sure how important it is that this gets turned into an assertion error. Either way.

Ashalynd replied:
It helps treating both Python versions in the same way.

Strilanc commented:
Please add the 're-review' tag when you've updated this, so it stands out.

Ashalynd commented on cirq/testing/order_tester.py:
```diff
@@ -0,0 +1,223 @@
+    def _try_comparison(self, comp):
+        result = NotImplemented
+        try:
+            result = comp()
```
This is done in such way to convert TypeErrors into AssertionErrors. It makes it a bit easier to use this class (only one type to catch), and is in line with what equals_tester is doing.  Also, it makes it uniform in both Python versions.
